src/app/OperationalDeviceProxy.h

```diff
@@ -72,7 +72,50 @@ struct DeviceProxyInitParams
 
 class OperationalDeviceProxy;
 
-typedef void (*OnDeviceConnected)(void * context, OperationalDeviceProxy * device);
+/**
+ * @brief Helper function as applications refactor away from holding onto a device proxy.
+ *
+ * Deprecated - Avoid using this object.
+ *
+ * This is an object that is meant to only be used by application that previously incorrectly held onto
+ * OperationDeviceProxy in previous implementations of OnDeviceConnected. It is expected that over time
+ * all instances of DeviceProxySession will be refactored out of existence.
+ */
+class DeviceProxySession : public DeviceProxy
+{
+public:
+    DeviceProxySession(Messaging::ExchangeManager * exchangeMgr, const SessionHandle & sessionHandle) :
+        mExchangeMgr(exchangeMgr), mSecureSession(sessionHandle), mPeerScopedNodeId(sessionHandle->GetPeer())
+    {}
+    DeviceProxySession() {}
+
+    void ShutdownSubscriptions() override { VerifyOrDie(false); } // Currently not implemented.
+    void Disconnect() override
+    {
+        // TODO should I be cleaning up mExchangeMgr, and mPeerScopedNodeId.
+        mSecureSession.Release();
```

Comment:
This isn't quite right - disconnection previously marked the session defunct.

Reply:
I was still trying to work out if we should be doing that if we can get the SessionHandle from the Holder. My argument against marking it as defunct comes from the multiple applications issue. If one application calls `Disconnect()` do we actually want to mark it as defunct?